cm2350/e200z7.py

```diff
@@ -108,115 +109,14 @@ def __init__(self, emu):
-        # already in the PpcRegOper class (and will be set in the i_mfspr()
-        # handler that calls this)
-        return self.systicks()
-
-    def tbuRead(self, emu, op):
-        '''
-        Read callback handler to associate the value of the TBU SPR with the
-        EmulationTime.
-        '''
-        # Get the top 32-bits of the "ticks" value.  This should be only 32-bits
-        # wide regardless of if this is a 64-bit or 32-bit machine.
-        return (self.systicks() >> 32) & 0xFFFFFFFF
-
-    def tblWrite(self, emu, op):
-        '''
-        Update the tb_offset so TBL values returned from this point on
-        reflect the new offset.
-        '''
-        # Ensure that the offset value is only 32-bits wide regardless of the
-        # platform size.
-        tbl_offset = self.getOperValue(op, 1) & 0xFFFFFFFF
-
-        # Based on the new TBL offset and the current value of TBU_WO, calculate
-        # the new desired timebase offset
-        tbu_offset = emu.getRegister(REG_TBU_WO)
-        offset = (tbu_offset << 32) | tbl_offset
-        self._tb_offset = super().systicks() - offset
-
-        # Return the offset so that TBL_WO has the correct offset to use to
-        # calculate the desired timebase offset.
-        return tbl_offset
-
-    def tbuWrite(self, emu, op):
-        '''
-        Update the tb_offset so TBU values returned from this point on
-        reflect the new offset.
-        '''
-        # Ensure that the offset value is only 32-bits wide regardless of the
-        # platform size.
-        tbu_offset = self.getOperValue(op, 1) & 0xFFFFFFFF
-
-        # Based on the new TBU offset and the current value of TBL_WO, calculate
-        # the new desired timebase offset
-        tbl_offset = emu.getRegister(REG_TBL_WO)
-        offset = (tbu_offset << 32) | tbl_offset
-        self._tb_offset = super().systicks() - offset
-
-        # Return the offset so that TBU_WO has the correct offset to use to
-        # calculate the desired timebase offset.
-        return tbu_offset
-
-    def systicks(self):
-        '''
-        Because PowerPC allows writes to the "Write-Only" TBL/TBU registers,
-        adjust the returned systicks value by the current offset.
-        '''
-        return super().systicks() - self._tb_offset
-
-
 import envi.archs.ppc.emu as eape
-import vivisect.impemu.platarch.ppc as vimp_ppc_emu
-#class PPC_e200z7(mmio.ComplexMemoryMap, eape.Ppc32EmbeddedEmulator, PpcEmulationTime):
-class PPC_e200z7(mmio.ComplexMemoryMap, vimp_ppc_emu.PpcWorkspaceEmulator, eape.Ppc32EmbeddedEmulator, PpcEmulationTime):
+import vivisect.impemu.emulator as vimp_emu
+#import vivisect.impemu.platarch.ppc as vimp_ppc_emu
+
+class PPC_e200z7(mmio.ComplexMemoryMap, vimp_emu.WorkspaceEmulator,
+                 eape.Ppc32EmbeddedEmulator, ppc_time.PpcEmuTime,
+                 #emutimers.ScaledEmuTimeCore, clocks.EmuClocks):
+                 emutimers.EmuTimeCore, clocks.EmuClocks):
```

Comment:
Are these related?  Should we really need to subclass them all here?  
The EmuClocks seems like an odd name for a subclassable class.  I was surprised to see it here.

Reply:
The `EmuClocks` module is a class that allows different peripherals to register a clock source that can then be retrieved in a generic way rather than requiring knowledge of which peripheral is providing the clock. This was added in [9091cdbf96ee7b8cc0e012a2fb4f2c9dfa2a34e4](https://github.com/Assured-Micropatching/CM2350-Emulator/commit/9091cdbf96ee7b8cc0e012a2fb4f2c9dfa2a34e4). A good example of this is in the `cm2350/peripherals/flexcan.py` file [line 1361](https://github.com/Assured-Micropatching/CM2350-Emulator/blame/54274f2be7debfe29dfbab9231ee3d42f32d69cd/cm2350/peripherals/flexcan.py#L1361)